File: src/main/java/com/tibiawiki/domain/utils/TemplateUtils.java
package com.tibiawiki.domain.utils;

import io.vavr.Tuple;
import io.vavr.Tuple2;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class TemplateUtils {

    private static final Logger log = LoggerFactory.getLogger(TemplateUtils.class);
    private static final String REGEX_PARAMETER_NEW = "\\|\\s+?([A-Za-z0-9_\\-]+)\\s*?=";
    private static final String REGEX_PARAMETER_LOWER_LEVELS = "\\|\\s+?lowerlevels\\s*?=((?:.*?\\{\\{.*?}})+)";
    private static final String REGEX_PARAMETER_LOWER_LEVELS_REMOVE = "\\|\\s+?lowerlevels\\s*?=((.*?\\{\\{.*?}})+)";
    private static final String LOWER_LEVELS = "lowerlevels";

    private TemplateUtils() {
        // no-args constructor, only static methods
    }

    public static String getBetweenOuterBalancedBrackets(String text, String start) {
        final Tuple2<Integer, Integer> startingAndEndingCurlyBrackets = getStartingAndEndingCurlyBrackets(text, start);
        return text.substring(startingAndEndingCurlyBrackets._1(), startingAndEndingCurlyBrackets._2());
    }

    /**
     * @param text  to search in
     * @param start string which denoted the start of the balanced brackets
     * @return two strings, the first is the substring of the provided text before the start of the balanced brackets,
     * the second is the substring after the start of the balanced brackets.
     */
    public static Tuple2<String, String> getBeforeAndAfterOuterBalancedBrackets(String text, String start) {
        final Tuple2<Integer, Integer> startingAndEndingCurlyBrackets = getStartingAndEndingCurlyBrackets(text, start);
        return Tuple.of(text.substring(0, startingAndEndingCurlyBrackets._1()), text.substring(startingAndEndingCurlyBrackets._2()));
    }

    /**
     * Remove the first line of the input string, that is, between the start of the string and the first occurrence
     * of a \n character.
     * Remove the last line of the input string, that is, everything after the last occurrence of two }} characters. The
     * reason we are not looking for the last \n character is that some templates may end with | notes =\n}} and then
     * removing the last \n character also removes the value of the key "notes".
     */
    @NotNull
    public static String removeFirstAndLastLine(@Nullable String text) {
        return Optional.ofNullable(text)
                .map(t -> t.substring(t.indexOf('\n') + 1)) // remove first line
                .map(t -> t.substring(0, t.lastIndexOf('\n') > -1 ? t.lastIndexOf("}}") : 0)) // remove last line
                .orElse("");
    }

    public static String removeStartAndEndOfTemplate(String text) {
        if (text.length() < 2) {
            return null;
        }
        int startOfTemplate = text.indexOf('|') + 1;
        int endOfTemplate = text.indexOf("}}");
        if (startOfTemplate >= 0 && endOfTemplate >= 0) {
            return text.substring(startOfTemplate, endOfTemplate).trim();
        }
        log.error("Could not remove start and end of template.");
        return null;
    }

    @NotNull
    public static Map<String, String> splitByParameter(@Nullable String infoboxTemplatePartOfArticle) {
        if (infoboxTemplatePartOfArticle == null || "".equals(infoboxTemplatePartOfArticle)) {
            return new HashMap<>();
        }

        final Map<String, String> keyValuePair = new HashMap<>();

        // first get keys
        List<String> keys = new LinkedList<>();
        Pattern pattern = Pattern.compile(REGEX_PARAMETER_NEW);
        Matcher matcher = pattern.matcher(infoboxTemplatePartOfArticle);
        while (matcher.find()) {
            if (matcher.groupCount() > 0 && matcher.group(1) != null) {
                String key = matcher.group(1);
                keys.add(key);
            }
        }

        final List<String> values = Arrays.asList((pattern).split(infoboxTemplatePartOfArticle));

        // sanitize values to get rid of empty Strings
        List<String> sanitizedValues = values.stream()
                .filter(s -> !s.isEmpty())
                .map(String::trim)
                .map(s -> s.replaceAll("\n$", ""))
                .map(s -> "".equals(s) ? null : s) // transform empty strings to null values
                .collect(Collectors.toList());

        if (keys.size() != sanitizedValues.size()) {
            if (log.isErrorEnabled()) {
                int endLength = infoboxTemplatePartOfArticle.length() >= 200 ? 200 : infoboxTemplatePartOfArticle.length();
                log.error("Amount of keys and values don't match for article starting with: {}",
                        infoboxTemplatePartOfArticle.substring(0, endLength).replaceAll("\\n", ""));
                return new HashMap<>();
            }
        }

        // put keys and values into map
        for (int i = 0; i < keys.size(); i++) {
            String key = keys.get(i);
            String value = sanitizedValues.get(i);
            keyValuePair.put(key, value);
        }

        return keyValuePair;
    }

    @NotNull
    public static List<String> splitByCommaAndTrim(@Nullable String input) {
        return Stream.of(input)
                .filter(Objects::nonNull)
                .filter(i -> i.trim().length() > 0)
                .map(i -> i.split(","))
                .flatMap(lst -> Arrays.stream(lst).map(String::trim))
                .collect(Collectors.toList());
    }

    @NotNull
    public static Optional<Map<String, String>> extractLowerLevels(@Nullable String infoboxTemplatePartOfArticleSanitized) {
        if (infoboxTemplatePartOfArticleSanitized == null || "".equals(infoboxTemplatePartOfArticleSanitized)) {
            return Optional.empty();
        }

        final Map<String, String> keyValuePair = new HashMap<>();

        Pattern pattern = Pattern.compile(REGEX_PARAMETER_LOWER_LEVELS, Pattern.DOTALL);
        Matcher matcher = pattern.matcher(infoboxTemplatePartOfArticleSanitized);
        while (matcher.find()) {
            if (matcher.groupCount() > 0 && matcher.group(1) != null) {
                String value = matcher.group(1);
                keyValuePair.put(LOWER_LEVELS, value);
            }
        }

        return keyValuePair.isEmpty()
                ? Optional.empty()
                : Optional.of(keyValuePair);
    }

    @NotNull
    public static String removeLowerLevels(@Nullable String infoboxTemplatePartOfArticleSanitized) {
        return Optional.ofNullable(infoboxTemplatePartOfArticleSanitized)
                .map(s -> Pattern.compile(REGEX_PARAMETER_LOWER_LEVELS_REMOVE, Pattern.DOTALL).matcher(s))
                .map(m -> m.replaceAll(""))
                .orElse("");
    }

    /**
     * @param text the text to search in
     * @param start the start String which denotes the start of the curly brackets
     * @return a tuple of two integers: the index of the start of the curly brackets and an index of the end of
     * the curly brackets
     */
    private static Tuple2<Integer, Integer> getStartingAndEndingCurlyBrackets(String text, String start) {
        final int startingCurlyBrackets = text.indexOf(start);

        if (startingCurlyBrackets < 0) {
            throw new IllegalArgumentException("Provided arguments text and start are not valid.");
        }

        int endingCurlyBrackets = 0;
        int openBracketsCounter = 0;
        char currentChar;

        for (int i = startingCurlyBrackets; i < text.length(); i++) {
            currentChar = text.charAt(i);
            if ('{' == currentChar) {
                openBracketsCounter++;
            }

            if ('}' == currentChar) {
                openBracketsCounter--;
            }

            if (openBracketsCounter == 0) {
                endingCurlyBrackets = i + 1;
                break;
            }
        }
        return Tuple.of(startingCurlyBrackets, endingCurlyBrackets);
    }
}

File: src/main/kotlin/com/tibiawiki/domain/utils/TemplateUtils.kt
package com.tibiawiki.domain.utils

import io.vavr.Tuple
import io.vavr.Tuple2
import org.slf4j.LoggerFactory
import java.util.Optional
import java.util.regex.Pattern

object TemplateUtils {
    private val log = LoggerFactory.getLogger(TemplateUtils::class.java)
    private const val REGEX_PARAMETER_INFOBOX_LINE = "\\|\\s+?([A-Za-z0-9_\\-]+)\\s*?="
    private const val REGEX_PARAMETER_LOOT_LINE = "\\|(.+?)\\s*?(=|,)"
    private const val REGEX_PARAMETER_LOWER_LEVELS = "\\|\\s+?lowerlevels\\s*?=((?:.*?\\{\\{.*?}})+)"
    private const val REGEX_PARAMETER_LOWER_LEVELS_REMOVE = "\\|\\s+?lowerlevels\\s*?=((.*?\\{\\{.*?}})+)"
    private const val LOWER_LEVELS = "lowerlevels"

    @JvmStatic
    fun getBetweenOuterBalancedBrackets(text: String, start: String): Optional<String> {
        return Optional.ofNullable(getStartingAndEndingCurlyBrackets(text, start))
            .map { text.substring(it._1(), it._2()) }
    }

    /**
     * @return two strings, the first is the substring of the provided text before the start of the balanced brackets,
     * the second is the substring after the start of the balanced brackets.
     */
    @JvmStatic
    fun getBeforeAndAfterOuterBalancedBrackets(text: String, start: String): Optional<Tuple2<String, String>> {
        return Optional.ofNullable(getStartingAndEndingCurlyBrackets(text, start))
            .map { Tuple.of(text.substring(0, it._1()), text.substring(it._2())) }
    }

    /**
     * Remove the first line of the input string, that is, between the start of the string and the first occurrence
     * of a \n character.
     * Remove the last line of the input string, that is, everything after the last occurrence of two }} characters.
     */
    @JvmStatic
    fun removeFirstAndLastLine(text: String?): String {
        return text
            ?.let { it.substring(it.indexOf('\n') + 1) }
            ?.let { t ->
                val lastNewline = t.lastIndexOf('\n')
                t.substring(0, if (lastNewline > -1) t.lastIndexOf("}}") else 0)
            }
            .orEmpty()
    }

    @JvmStatic
    fun removeStartAndEndOfTemplate(text: String): String? {
        if (text.length < 2) {
            return null
        }
        val startOfTemplate = text.indexOf('|') + 1
        val endOfTemplate = text.indexOf("}}")
        if (endOfTemplate >= 0) {
            return text.substring(startOfTemplate, endOfTemplate).trim()
        }
        log.error("Could not remove start and end of template.")
        return null
    }

    @JvmStatic
    fun splitInfoboxByParameter(infoboxTemplatePartOfArticle: String?): MutableMap<String, String?> {
        return splitByParameter(infoboxTemplatePartOfArticle, REGEX_PARAMETER_INFOBOX_LINE)
    }

    @JvmStatic
    fun splitLootByParameter(lootTemplatePartOfArticle: String?): MutableMap<String, String?> {
        return splitByParameter(lootTemplatePartOfArticle, REGEX_PARAMETER_LOOT_LINE)
    }

    @JvmStatic
    fun splitByParameter(infoboxTemplatePartOfArticle: String?, regex: String): MutableMap<String, String?> {
        if (infoboxTemplatePartOfArticle.isNullOrEmpty()) {
            return HashMap()
        }

        val keyValuePair = HashMap<String, String?>()
        val keys = mutableListOf<String>()
        val pattern = Pattern.compile(regex)
        val matcher = pattern.matcher(infoboxTemplatePartOfArticle)
        while (matcher.find()) {
            if (matcher.groupCount() > 0 && matcher.group(1) != null) {
                keys.add(matcher.group(1))
            }
        }

        val values = pattern.split(infoboxTemplatePartOfArticle).toList()
        val sanitizedValues = values
            .filter { it.isNotEmpty() }
            .map { it.trim() }
            .map { it.replace(Regex("\n$"), "") }
            .map { if (it.isEmpty()) null else it }

        if (keys.size != sanitizedValues.size && log.isErrorEnabled) {
            val endLength = minOf(infoboxTemplatePartOfArticle.length, 200)
            log.error(
                "Amount of keys and values don't match for article starting with: {}",
                infoboxTemplatePartOfArticle.substring(0, endLength).replace("\n", "")
            )
            return HashMap()
        }

        for (i in keys.indices) {
            keyValuePair[keys[i]] = sanitizedValues[i]
        }

        return keyValuePair
    }

    @JvmStatic
    fun splitByCommaAndTrim(input: String?): List<String> {
        return sequenceOf(input)
            .filterNotNull()
            .filter { it.trim().isNotEmpty() }
            .flatMap { it.split(",").asSequence().map(String::trim) }
            .toList()
    }

    @JvmStatic
    fun extractLowerLevels(infoboxTemplatePartOfArticleSanitized: String?): Optional<Map<String, String>> {
        if (infoboxTemplatePartOfArticleSanitized.isNullOrEmpty()) {
            return Optional.empty()
        }

        val keyValuePair = HashMap<String, String>()
        val pattern = Pattern.compile(REGEX_PARAMETER_LOWER_LEVELS, Pattern.DOTALL)
        val matcher = pattern.matcher(infoboxTemplatePartOfArticleSanitized)
        while (matcher.find()) {
            if (matcher.groupCount() > 0 && matcher.group(1) != null) {
                keyValuePair[LOWER_LEVELS] = matcher.group(1)
            }
        }

        return if (keyValuePair.isEmpty()) {
            Optional.empty()
        } else {
            Optional.of(keyValuePair)
        }
    }

    @JvmStatic
    fun removeLowerLevels(infoboxTemplatePartOfArticleSanitized: String?): String {
        return infoboxTemplatePartOfArticleSanitized
            ?.let { Pattern.compile(REGEX_PARAMETER_LOWER_LEVELS_REMOVE, Pattern.DOTALL).matcher(it) }
            ?.replaceAll("")
            .orEmpty()
    }

    private fun getStartingAndEndingCurlyBrackets(text: String, start: String): Tuple2<Int, Int>? {
        val startingCurlyBrackets = text.indexOf(start)
        if (startingCurlyBrackets < 0) {
            return null
        }

        var endingCurlyBrackets = 0
        var openBracketsCounter = 0
        for (i in startingCurlyBrackets until text.length) {
            val currentChar = text[i]
            if (currentChar == '{') {
                openBracketsCounter++
            }
            if (currentChar == '}') {
                openBracketsCounter--
            }
            if (openBracketsCounter == 0) {
                endingCurlyBrackets = i + 1
                break
            }
        }
        return Tuple.of(startingCurlyBrackets, endingCurlyBrackets)
    }
}
